### ISSUE TITLE
[REQ] Destruct tasks after they finish

### ISSUE THREAD

**Issue by mokafolio** [bshoshany/thread-pool#81]
**Describe the new feature**

The way the worker function is implemented right now, tasks are not destructed until a worker starts a new task or the thread pool de-initializes. While this is not a problem in many cases, if your submitted function captures objects where lifetime is important this can be rather confusing.

**Code example**

An example of code that utilizes the suggested feature. Paste or write it between the two code fences.

```cpp
struct Foo
{
    ~Foo
    {
        printf("I Died\n");
    }
};

int main(int _argc, const char * _args[])
{
    thread_pool_light tp;
    {
        auto shared_foo = std::make_shared<Foo>();
        auto fut = tp.submit([shared_foo]() {});
    }
    //...do some other stuff

    //...Foo won'd destruct until the threadpool goes out of scope even though the task is done.
}
```

**Comment by bshoshany** [comment]
Thanks for the suggestion! However, I believe the functionality you want is actually already implemented. If you run your example code, you will see that the object is destructed before the pool goes out of scope. Here's a working example:

```cpp
#include "BS_thread_pool_light.hpp"
#include <iostream>

struct Foo
{
    ~Foo()
    {
        printf("I Died\n");
    }
};

int main()
{
    BS::thread_pool_light tp;
    {
        auto shared_foo = std::make_shared<Foo>();
        auto fut = tp.submit([shared_foo]() {});
    }
    std::cout << "Thread pool still in scope...\n";
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
    std::cout << "Thread pool will now go out of scope.\n";
}
```

When I run this (GCC 12.2.0) I get the following output:

```none
Thread pool still in scope...
I Died
Thread pool will now go out of scope.
```

This means that `Foo` is destroyed once the task is done, **not** when the thread pool goes out of scope. Am I missing something here?

**Comment by mokafolio** [comment]
Sorry about that, I think my initial issue must have been caused by something else. For some reason putting the task into an extra block inside the worker fixed it. I consequently tunnel visioned and brain farted. Sorry!
